src/GasEstimator.ts

```diff
@@ -0,0 +1,262 @@
+        }
+    );
+
+    internalEstimation = applyInternalCorrection(internalEstimation);
+
+    const relayEstimation = estimateMaxPossibleRelayCallWithLinearFit(
+        internalEstimation.toNumber(),
+        tokenEstimation.toNumber()
+    );
+
+    return BigNumber(relayEstimation);
+};
+
+/**
+ * Estimates the max possible gas consumed by transfering an ERC20 token
+ * @param contractInteractor object containing the contractInteractor
+ * @param request request that contains the relayRequest/deployRequest and metadata
+ * @returns gas estimation from the transfer
+ */
+export const estimateMaxPossibleGasTokenTransfer = async (
+    contractInteractor: ContractInteractor,
+    { request, relayData }: RelayRequest | DeployRequest
+): Promise<BigNumber> => {
+    const deploy = isDeployRequest(request);
+    let tokenEstimation: BigNumber;
+    const tokenGas = BigNumber(request.tokenGas);
+    if (tokenGas.gt(0)) {
+        tokenEstimation = BigNumber(request.tokenGas);
+    } else {
+        const erc20: ERC20Token = await contractInteractor.getERC20Token(
+            request.tokenContract
+        );
+
+        const methodToEstimate = await erc20.instance.contract.methods.transfer(
+            relayData.feesReceiver,
+            toWei('1')
+        );
+
+        let caller = relayData.callForwarder;
+        if (deploy) {
+            const { from, recoverer, index } = request as DeployRequestStruct;
+            caller = await contractInteractor.getSmartWalletAddress(
+                caller,
+                from,
+                recoverer,
+                index
+            );
+        }
+
+        const estimation = await methodToEstimate.estimateGas({
+            from: caller,
+            gasPrice: relayData.gasPrice
+        });
+
+        tokenEstimation = applyInternalCorrection(estimation);
+    }
+
+    if (tokenEstimation.isZero() && deploy) {
+        return SUBSIDY;
+    }
+
+    return applyGasCorrectionFactor(tokenEstimation);
+};
+
+/**
+ * Estimates the max possible gas consumed by executing a function
+ * @param contractInteractor object containing the contractInteractor
+ * @param estimate params to execute the estimateGas function
+ * @returns gas estimation from the execution
+ */
+export const estimateMaxPossibleGasExecution = async (
+    contractInteractor: ContractInteractor,
+    estimate: EstimateGasParams
+): Promise<BigNumber> => {
+    const estimation = await contractInteractor.estimateGas(estimate);
+    return BigNumber(estimation);
+};
+
+/**
+ * Applies the correction from internal calls
+ * @param estimation BigNumber gas estimation that needs to be corrected
+ * @returns gas estimation with the correction done
+ */
+export const applyInternalCorrection = (
+    estimation: BigNumber | number | string
+): BigNumber => {
+    const bigValue = BigNumber(estimation);
+    const correction = BigNumber(
+        constants.INTERNAL_TRANSACTION_ESTIMATE_CORRECTION
```

Comment:
another annoying, out-of-scope question: why does this value come from common? surely this is a configuration property, isn't it? If so it should be defined in the client.configuration perhaps?

Reply:
Im not sure, but we need to decide where to put the configurations and identify if the client can be a dependency from the server.

Reply:
Not sure whether this value should be set in the client or not, however I do agree it's out of scope 😅